bpo-37903: IDLE: Shell sidebar with prompts

taleinat opened the pull request:
This is a working implementation of a sidebar for IDLE's shell, removing the prompts from the main text widget (except for special prompts such as the one used by the debugger.) It works very well in my testing.

__Please give this a try and let me know what you think!__

Currently known to be missing:
1. handling of left mouse button clicks/drags on the sidebar

Notes on the implementation:
* I mostly made each change in a separate commit to make reviewing easier.
* This does *not* use the same base-class as the editor line-numbers sidebar, because it uses a canvas widget rather than a text widget. This is needed due to the shell soft-wrapping lines, causing lines to have varied height. Also, Squeezer inserts buttons into the shell's text window, and those lines have a different height as well.
* This includes quite a bit of work to get the "stdin" tag applied consistently to user input, which was rather broken but not relied on for anything. (Actually, this fixes the "recall" functionality in some edge cases.)
* Removing prompts from the text widget made knowing where each statement started impossible, but the sidebar needs to know this. This PR adds the "console" tag to the newline before each statement to overcome this.
* Removing prompts has other side effects which needed to be addressed. For example, successive statements without output now create a single long range of input tagged "stdin", which required some changes to the `recall()` method.

<!-- issue-number: [bpo-37903](https://bugs.python.org/issue37903) -->
https://bugs.python.org/issue37903
<!-- /issue-number -->

taleinat committed:
initial working shell sidebar

taleinat committed:
fix recall and handling of prompt and input lines

taleinat committed:
update only as needed during editing, accounting for line wrapping

taleinat committed:
properly insert delegators into their proper place in the percolator

Specifically, avoid EditorWindow.ResetColorizer() moving the undo
and colorizer delegators to the top of the percolator, by removing
and inserting them.

taleinat committed:
move non-syntax tag highlighting out of ColorDelegator

This avoids ColorDelegator removing unrelated tags, specifically "stdin",
when re-colorizing.

taleinat committed:
replace shell sidebar prompt and input line recognition method

Prompts are now marked in the text by adding the "console" tag to
the newline character ('\n') before the beginning of the line.  This
works around issues with the prompt itself usually no longer being
written in the text widget.

With another small fix to the PyShell.recall() method, input lines
can now be recognized consistently by checking for the "stdin" tag.

This is much better than attempting to separately track the prompt
and input lines, since those are difficult to keep up to date after
undo/redo actions and squeezing/unsqueezing of previous output.

taleinat committed:
minor code cleanup

taleinat committed:
remove unnecessary .replace() delegator method

taleinat committed:
fix inverted condition in deciding whether to update sidebar (oops!)

taleinat committed:
fix "stdin" tag removed when undo-ing a deletion

taleinat committed:
fix sidebar update on new prompt

taleinat committed:
add a NEWS entry

taleinat committed:
fix recall handling with multiple successive statements

taleinat committed:
remove a dead line

taleinat committed:
remove commented out initial attempt

aeros reviewed (commented):
Thanks for the PR @taleinat. I'm definitely onboard with the concept of this change, a dedicated sidebar prompt for the shell would be a nice addition.

From an initial review of the GUI, it looks like the font of the characters in the sidebar aren't scaling dynamically when the font is adjusted in the shell:

Opening the shell with a font size of 22 Arial (expected):
![image](https://user-images.githubusercontent.com/44193521/63643747-8da7d200-c6a5-11e9-95a6-f3836f80c5ad.png)

Opening the shell while the font size is set to 8 Arial and then adjusting it to 22 within the shell interface (Not expected):
![image](https://user-images.githubusercontent.com/44193521/63643753-bb8d1680-c6a5-11e9-8eba-c632bc815443.png)

taleinat committed:
update shell sidebar font upon font config changes

terryjreedy commented on Lib/idlelib/colorizer.py:
```diff
@@ -120,12 +120,10 @@ def LoadTagDefs(self):
-            # The following is used by ReplaceDialog:
-            "hit": idleConf.GetHighlight(theme, "hit"),
-            }
+        }
```
Unlike tab indents, these changes, in particular the movement of 'hit' retrieval don't seem even indirectly related to the sidebar.  I can't see the net changes as an improvement.

taleinat replied:
See my other comment, regarding `get_tag_colors()`, explaining this.

terryjreedy replied:
I completely agree that colorizer should not delete tags that are needed but not restored by colorizer.  But I don't think that this ever happens with 'hit' as it is only used while modal dialogs are open.  So no fix is needed for 'hit' now.

The irony of this discussion about the 'hit' (found) tag is that, on Windows at least, it is effectively unused by ReplaceDialog since hits are first tagged with the select tag so that the hit tag is ignored.  (At least on Windows, debugger uses it instead of the selection tag as Windows does not allow selections in inactive windows.  )  I think it should be used and visible in find, replace, and grep, but this is obviously another issue. 

As long as 'hit' is not used by pyshell, it has nothing to do with this issue and partially fixing it, but not really  fixing it, is a distraction.  So let's delete the 'hit' changes.  If we really fix 'hit', we might treat it separately, perhaps in searchbase, much as 'breakpoint' is.  The most we might do here is add a TODO comment here about moving it someday so that colorizer does not delete it, should that become possible.

I should have suggested two weeks ago that the minimal coloring fixes that *are*needed for Shell be split off into a separate PR to be discussed and merged first.  Too late now, but it is a somewhat separate discussion.

taleinat replied:
Since this PR requires the removal of the shell tags from the colorizer's `tag_defs`, and the refactoring that entails, removing "hit" which obviously shouldn't be there was a simple and obvious fix. I'll revert it with a comment if you prefer.

(I will mention that this will affect my PR replacing the search/replace dialogs with a toolbar: GH-9855).

terryjreedy replied:
I left a comment on your toolbar issue.  With the comment added and 'hit' left, we can resolve this for now.

terryjreedy commented on Lib/idlelib/editor.py:
```diff
@@ -790,18 +788,40 @@ def _rmcolorizer(self):
         if self.line_numbers is not None:
             self.line_numbers.update_colors()
 
+    def get_tag_colors(self):
```
This confuses me. The name is misleading, as it should be 'get_hit_colors'.  It is a one-line function used only twice that for some reason singles out only one of the theme elements.  It is not a method (we need to discuss this point), and if it is be extracted, it should be a function defined in config after idleConf.

aeros replied:
@terryjreedy:
>  It is not a method (we need to discuss this point), and if it is be extracted, it should be a function defined in config after idleConf.

I'm not sure that I understand what you mean here. Do you mean that it shouldn't be a method? 

From my understanding, in the current implementation @taleinat is using, ``get_tag_colors`` would definitely be considered a method, as it contains a ``self`` parameter and is being used within the ``EditorWindow`` class.

I don't mean to nitpick at all here, I was just a bit confused and wanted to make sure that I'm not misunderstanding what you meant.

taleinat replied:
Note that PyShell extends this method.

This is intended as a way for sub-classes to add tag coloring definitions, and PyShell uses it to define colorings for the "console", "stdout", "stderr" and "stdin" tags.

In the existing codebase (before this PR), these (and "hit") were shoved into ColorDelegator, which had the unfortunate side-effect of ColorDelegator clearing these tags whenever it re-colorized a block of code. This is a bug, resulting from mixing tags used only for syntax highlighting (which can be safely cleared and set again later) with other tags.

This PR needs this bug fixed to keep these tags in place, especially the "console" and "stdin" tags which it uses to recognize prompts and user inputs.

terryjreedy replied:
@aeros167 The self parameter is not used, which means it could be removed.  It could be retained as a staticmethod but I don't believe in that (another discussion for another time.

terryjreedy replied:
If we needed to do anything about 'hit' (which we don't), I would prefer to put one combined line where the two calls are.  PyShell does not use this method for the PyShell tags.  The super call in the PyShell code depends on a broken inheritance structure that I intend to change.

taleinat replied:
I've removed the `get_tag_colors()` method.

terryjreedy commented on Lib/idlelib/percolator.py:
```diff
@@ -38,6 +38,21 @@ def insertfilter(self, filter):
         self.top = filter
 
+    def insertfilterafter(self, filter, after):
+        assert isinstance(filter, Delegator)
```
We do not routinely use assert.  Why here?  The lack of docstring and tests?  Bad design such that bad calls do not otherwise fail?  If filter is not a Delegator, would not filter.setdelegate fail?  If so, no check is needed here.

The advice I have seen is that assert, which can be optimized away, should only be used to check internal logic.  I guess the question is where the boundary of internal for IDLE.  If we do use assert, we should make sure it will be executed during testing.

taleinat replied:
> We do not routinely use assert. Why here? ...

I simply mirrored the existing use in `Percolator.insertfilter()`, to which this is very similar.

terryjreedy replied:
See next response.

terryjreedy commented on Lib/idlelib/percolator.py:
```diff
@@ -38,6 +38,21 @@ def insertfilter(self, filter):
+
+        f = self.top
+        f.resetcache()
+        while f is not after:
```
Assert strikes me as an odd (confusing) way to end a loop.  More normal:
```suggestion
        while f is not after and f is not self.bottom:
```
followed after the loop by
        if f is self.bottom: raise ValueError("after must be in the delegator chain")

taleinat replied:
Note that this mirrors the very similar piece of logic in `Percolator.removefilter()`.

The assert isn't ending the loop. The code is kept simpler by assuming that `after` is in the percolator; this simply validates that. My understanding is that this is done in hope that this will catch bugs early enough in the development process.

I agree, more generally, that Percolator should be made safer and better tested, or removed entirely in favor of refactoring to a less confusing design. However, given its current state, I believe that this change is in line with the "status quo".

terryjreedy replied:
It is hard to review diffs to code that needs an overhaul.  So leave what you did until then.

terryjreedy commented on Lib/idlelib/editor.py:
```diff
@@ -790,18 +788,40 @@ def _rmcolorizer(self):
+        """
         EditorWindow.color_config(self.text)
 
+        tag_colors = self.get_tag_colors()
```
If editor needs the 'hit' colors, why not add it to self.color_config?

taleinat replied:
See my other comment, regarding `get_tag_colors()`, explaining this.

terryjreedy replied:
A better answer, which I had forgotten, is that color_config sets text color options,  and only those, and 'hit' has nothing to do with them.  It has nothing to do with this issue.

terryjreedy commented on Lib/idlelib/pyshell.py:
```diff
@@ -380,6 +366,27 @@ def delete(self, index1, index2=None):
             pass
         UndoDelegator.delete(self, index1, index2)
 
+    def undo_event(self, event):
```
What would be the manual test of this change?

taleinat replied:
In the shell, write some code, delete some of it, then undo the deletion. The "stdin" tag should be present on the restored input.

terryjreedy replied:
The 'stdin' is a null tag that currently has no effect.  So its presence is invisible.  This patch appears to change the stdin background so that it is not longer the default text background which is the normal text background.  But I do not yet see how this happens.  This is the 'background glitch' I mentioned.

terryjreedy commented on Lib/idlelib/pyshell.py:
```diff
@@ -334,34 +335,19 @@ def open_shell(self, event=None):
         self.tag_add("SYNC", "1.0", "iomark")
         ColorDelegator.recolorize_main(self)
 
-    def LoadTagDefs(self):
```
If I understand correctly, you moved the non-code colors from this ColorDelegator subclass to the Shell editor subclass because the colorizer does not use them and need not know about them.

taleinat replied:
See my other comment, regarding `get_tag_colors()`, explaining this.

In a sentence, this is indeed proper separation of concerns, but more importantly this fixes a nasty bug.

terryjreedy commented on Lib/idlelib/pyshell.py:
```diff
@@ -929,6 +937,33 @@ def __init__(self, flist=None):
+            "stdout": idleConf.GetHighlight(theme, "stdout"),
+            "stderr": idleConf.GetHighlight(theme, "stderr"),
+            "console": idleConf.GetHighlight(theme, "console"),
+        })
```
It would be clearer to just add a line at the end:
`            "hit": idleConf.GetHightlight(theme, "hit"),

taleinat replied:
You mean overriding the parent class's method rather than extending it? Isn't that a recipe for future bugs in later changes to `EditorWindow` or other classes in the hierarchy?

terryjreedy replied:
The parent class of PyShell should not have Editor-only specializations that have to be undone by subclasses.  Depending of this broken structure is also a recipe for future bugs ;-).  Anyway, this is a moot point if we leave 'hit' alone.

terryjreedy commented on Lib/idlelib/pyshell.py:
```diff
@@ -1163,13 +1198,30 @@ def enter_callback(self, event):
         # trailing whitespace
         if self.text.compare("insert", "<", "iomark linestart"):
-            # Check if there's a relevant stdin range -- if so, use it
+            # Check if there's a relevant stdin range -- if so, use it.
```
I skipped this block for now.

terryjreedy commented on Lib/idlelib/pyshell.py:
```diff
@@ -1222,15 +1274,16 @@ def recall(self, s, event):
                 self.newline_and_indent_event(event)
                 prefix = self.text.get("insert linestart", "insert")
-            self.text.insert("insert", lines[0].strip())
+            self.text.insert("insert", lines[0].strip(), "stdin")
```
I believe the addition of 'stdin' tags has something to do with the background change I mentioned in point 3.  But I am not concerned about it now.

terryjreedy commented on Lib/idlelib/pyshell.py:
```diff
@@ -1273,7 +1330,14 @@ def restart_shell(self, event=None):
+
+        prompt = self.prompt
+        if self.sys_ps1 and prompt.endswith(self.sys_ps1):
+            prompt = prompt[:-len(self.sys_ps1)]
```
This is pretty useless.  PyShell is initialized with
```
        self.sys_ps1 = sys.ps1 if hasattr(sys, 'ps1') else '>>> '
        self.prompt_last_line = self.sys_ps1.split('\n')[-1]
        self.prompt = self.sys_ps1  # Changes when debug active
```
The only way I know to either reset or delete sys.ps1 and start IDLE in the same process is to do so in python code and then `import idlelib.idle`.  Under normal circumstances, self.sys_ps1 exists, is not '', and and after the following code, equals self.prompt and self.prompt_last_line.

The new conditional will trigger once and clear the prompt to '' (or remove it from 'DEBUG ON\n'.   The conditional will never trigger again.  After it is cleared, the only possible 'prompt' within the text box is "DEBUG ON\n" .  (I don't think it should be a repeated prompt -- another issue.)  So we should remove sys_ps1.  Until we remove the debug prompt, we should just initialize self.prompt it to ''.

self.prompt_last_line is more problematical.  It should be set to '' or better deleted, but it is used (only) in EditorWindow to indicate "isinstance(self, PyShell)" and affect backspace and HOME.  (The assumption is that it is never '' in Shell, but it will be if sys.ps1 ends with '\n'.)  The uses of prompt_last_line need to be reviewed to remove uses of the string as a string and replace it with a boolean where the indicator is still needed.  (I started on this.)

taleinat replied:
Indeed, this is needed to support custom prompts, as currently used only by the debugger (and possibly by extensions).

I was aiming to change as little as possible. The places where `prompt_last_line` is used are rather delicate, and changing them could requiring making other related changes... @terryjreedy, are you sure you want this as part of this PR?

terryjreedy replied:
Setting self.prompt directly to '' is less change that setting it to '>>>' and then correcting it to ''.  Either way, the debugger will add and delete its notice.

prompt_last_line is a separate issue.  Giving it a now incorrect value is what strikes me as dangerous.  Maybe it never matters, maybe it will in certain situations.

terryjreedy commented on Lib/idlelib/sidebar.py:
```diff
@@ -1,19 +1,33 @@
 def get_end_linenumber(text):
     """Utility to get the last line's number in a Tk text widget."""
-    return int(float(text.index('end-1c')))
+    return get_lineno(text, 'end-1c')
```
The docstrings violate pep8 and only 1 function is needed.
```
def get_lineno(text, idex='end-1c'):
    "Return line number in text of idex, which default to the last line"
    return int(float(text.index(idex)))

taleinat replied:
I guess it's a matter of style, but IMO having `get_lineno()` return the last line number by default isn't an intuitive default behavior. It feels like favoring conciseness over clarity, which I usually try to avoid.

aeros replied:
@terryjreedy:
> The docstrings violate pep8

As far as I can tell, this would not be in violation of [PEP-257's section on one line docstrings](https://www.python.org/dev/peps/pep-0257/#one-line-docstrings) though. PEP-257 is main authority on docstrings. If anything, it indicates that triple quotes are preferable over single quotes:

> Triple quotes are used even though the string fits on one line. This makes it easy to later expand it.

terryjreedy replied:
PEP-257 says that docstrings should begin with a command verb, usually 'Return', occasionally something else.

`"Return x"` is common in the stdlib.  `"""Return x"""` takes 4 extra spaces, which might be needed, and to me is too heavy for a single line.

`get_end_linenumber(text)` is only 5 chars shorter than `get_lineno(text, 'end-1c')`.  Hardly enough to justify a new function called 3 times.  How about adding `END =  'end-1c'` and making the calls  `get_lineno(text, END)`?

taleinat replied:
This module currently consistently uses triple-quotes even for single-line doc-strings.

I'd be happy to change that in a follow-up PR, but I think it's better to avoid making more widespread changes in this (already large) PR, and I wouldn't want to break the current consistency.

terryjreedy replied:
Leave the triple quote if you prefer and space allows.  I am more concerned about an unneeded quarter-liner.

terryjreedy commented on Lib/idlelib/sidebar.py:
```diff
@@ -40,6 +54,15 @@ def get_widget_padding(widget):
     return padx, pady
 
 
+@contextlib.contextmanager
```
This or something like it would seem useful in pyshell too.  Yet another spinoff.

taleinat replied:
Indeed!

terryjreedy commented on Lib/idlelib/sidebar.py:
```diff
@@ -40,6 +54,15 @@ def get_widget_padding(widget):
+
 class BaseSideBar:
     """
     The base class for extensions which require a sidebar.
```
'extensions is not the right word anymore.

I am stopping here for now.

aeros replied:
@terryjreedy:
> 'extensions is not the right word anymore.

Would the correct term here be "derived class"? That seems to be the term used in https://docs.python.org/3.9/tutorial/classes.html, but there may a better word to use that would be more specific to this context. 

If "extensions" is no longer the correct term to use, the documentation for IDLE should probably be updated accordingly https://docs.python.org/3.9/library/idle.html?highlight=extensions. Unless the documentation is referring to something different from what the code comment is referring to.

Also, minor grammar nit: "that" is more correct than "which" in this context. 

```suggestion
     The base class for derived classes that require a sidebar.
```

terryjreedy replied:
See `idlelib/config_extensions.def` for the meaning of 'extensions'.  Many IDLE features were once implemented as extensions but no longer.

terryjreedy commented on Lib/idlelib/sidebar.py:
```diff
@@ -40,6 +54,15 @@ def get_widget_padding(widget):
+
 class BaseSideBar:
     """
     The base class for extensions which require a sidebar.
```
Docstrings should not start with quotes-only on the first line.  Delete both and 
```suggestion
    "A base class for sidebars using Text."
```

taleinat replied:
Good point.

taleinat committed:
update sidebar text colors upon highlight config changes

taleinat committed:
use the prompt ("console") foreground color for the shell sidebar text

The background color still uses that configured for line numbers.

taleinat committed:
show continuation prompts ("...") for multi-line history recall

The fix needed was for History to properly set the "stdin" tag
on text it inserts.

taleinat committed:
show continuation prompts ("...") after multi-line replace in the shell

taleinat committed:
show continuation prompts ("...") after multi-line undo/redo in the shell

taleinat committed:
don't show continuation prompts ("...") for print() outputs

taleinat committed:
allow deleting ">>> " at the beginning of a line in shell windows

taleinat committed:
update shell sidebar on shell output squeeze/unsqueeze

taleinat committed:
remove failing editor window test that is no longer relevant

taleinat committed:
add several tests for the shell sidebar

taleinat committed:
reduce shell sidebar test run time by reusing the PyShell

terryjreedy commented on Lib/idlelib/idle_test/test_sidebar.py:
```diff
@@ -370,5 +378,164 @@ def assert_colors_are_equal(colors):
+        idlelib.pyshell.use_subprocess = True
+
+        cls.root = root = tk.Tk()
+
```
```suggestion
        root.withdraw()
```
Tests still pass without extra window.

taleinat replied:
Good catch!

terryjreedy commented on Lib/idlelib/idle_test/test_sidebar.py:
```diff
@@ -370,5 +378,164 @@ def assert_colors_are_equal(colors):
         assert_colors_are_equal(orig_colors)
 
 
+class TestShellSidebar(unittest.TestCase):
```
'IDLE Convention', Test at end:
```suggestion
class ShellSidebarTest(unittest.TestCase):
```

terryjreedy commented on Lib/idlelib/idle_test/test_sidebar.py:
```diff
@@ -370,5 +378,164 @@ def assert_colors_are_equal(colors):
+            expected_lines,
+        )
+
+    def getShellLineYCoords(self):
```
Except for assert methods, please stick with PEP8 and current IDLE convention for function names.
```suggestion
    def get_shellline_ycoords(self):
```
Etcetera

taleinat replied:
Bah, sorry, silly mistake!

terryjreedy commented on Lib/idlelib/idle_test/test_sidebar.py:
```diff
@@ -370,5 +378,164 @@ def assert_colors_are_equal(colors):
+        self.assertSidebarLinesEndWith(['>>>', '   ', '>>>'])
+
+    def testMultiLineCommand(self):
+        # note: block statements are not indented because IDLE auto-indents
```
```suggestion
        # Block statements are not indented here because IDLE auto-indents.
```
PEP8 comments.

taleinat replied:
Done.

terryjreedy commented on Lib/idlelib/idle_test/test_sidebar.py:
```diff
@@ -1,4 +1,7 @@
 """Test sidebar, coverage 93%"""
```
```suggestion
"""Test sidebar, coverage 84%"""
```
On my machine it was 91% today before patch.

taleinat replied:
I hope to increase coverage before this is merged by adding more tests. So let's not make this change quite yet.

terryjreedy replied:
Fine.  We can leave this unresolved as a reminder.

taleinat replied:
I've updated this now that I'm done removing dead code and adding tests (now at 96%).

terryjreedy reviewed (changes requested):
Of course I would like a faster test.  Some sleeps may be unneeded.  Some are not always enough.  Frustrating. though we don't know if failures will occur on buildbots.  

By running test_sidebar 20 times while running the Python test suite in parallel, thus using all cores, I got one failure of testSqueezeSingleLineCommand.

I would am not sure what you mean by 'general approach'.  What would be a different approach?

Comments other than test_sidebar are leftover from some time past and possibly repeats.

taleinat committed:
revert removal of the "hit" tag from Colorizer.tag_defs

taleinat committed:
hide unnecessary extra "root" Tk window in tests

taleinat committed:
fix test method names

taleinat committed:
fix test comments according to pep-8 style

taleinat committed:
rename sidebar test class according to convention

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

taleinat committed:
re-word doc-strings for BaseSideBar and EndLineDelegator

taleinat committed:
re-word doc-strings for get_lineno and get_end_linenumber

taleinat committed:
optimize tests and make them more robust

taleinat committed:
handle mouse buttons except for button 1 (left-click)

taleinat committed:
add tests for text font and color configs

terryjreedy commented on Lib/idlelib/squeezer.py:
```diff
@@ -285,7 +287,7 @@ def count_lines(self, s):
 
-    def squeeze_current_text_event(self, event):
+    def squeeze_current_text(self):
         """squeeze-current-text event handler
```
Start with verb, end line with period.
```suggestion
        """Handle <<squeeze-current-text>> pseudoevent.
```
Brackets could be left off.

taleinat replied:
Done.

taleinat committed:
remove show/hide_sidebar from the shell sidebar

taleinat committed:
add more test cases

taleinat committed:
update coverage percentage

taleinat committed:
re-word squeeze_current_text() doc-string

taleinat committed:
update coverage percentage
